First draft of a blog post about the Spark streaming dataset

stichbury opened the pull request:
Here's a first draft ready for the team to improve.

tingtingQB commented on blog_post_collaboration/spark_streaming_dataset.md:
Any kedro-datasets version after 1.4.0, 
e.g., `pip install “kedro-datasets==1.4.0”`

tingtingQB replied:
Should be `kedro==0.18.9?`?

tingtingQB commented on blog_post_collaboration/spark_streaming_dataset.md:
Once we installed the latest version of kedro-datasets, this step can be skipped

tingtingQB commented on blog_post_collaboration/spark_streaming_dataset.md:
This post is to give data practitioners a guidance on how to use streaming datasets, 
Limitations:
1. Not ready for service broker e.g., Kafka, as the additional Jar package is required
2. Schema is enforced

tingtingQB reviewed (commented):
I have added some comments, please can you also added Tom Kurian as reviewer?

tingtingQB commented on blog_post_collaboration/spark_streaming_dataset.md:
We had a look on the original issue with YARN cluster mode, I think it is no more the case, because in this latest article https://kedro.org/blog/how-to-deploy-kedro-pipelines-on-amazon-emr , we mentioned how to deploy kedro pipeline on another application cluster.
In order to run spark streaming with the production workflow, as long as we spark-submit a job to have a running cluster with proper scheduler, it should work, an client project example was we used streaming on Databrick clusters.

stichbury replied:
That's really useful, thanks @tingtingQB. I've made a few edits here to the summary but could use a little extra help from you, Haris or Tom to refine what i've added, and also to smooth out the introduction and summary to add any further points you think a reader would find useful.

stichbury replied:
cc https://github.com/kuriantom369 & https://github.com/harisqb

astrojuanlu replied:
cc @kuriantom369 & @harisqb (full urls don't work as mentions I think)

tingtingQB reviewed (commented):
Additional comment added

harisqb commented on blog_post_collaboration/spark_streaming_dataset.md:
I think I would move all this in the introduction, might get lost at the bottom of the page

stichbury commented on blog_post_collaboration/spark_streaming_dataset.md:
```diff
@@ -0,0 +1,322 @@
+
+Be aware that windowing operations only allow windowing on time columns.
+
+Watermarks must be defined for joins. Only certain types of joins are allowed, and these depend on the file types (stream-stream, stream-static) which makes joining of multiple tables a little complex at times. For further information or advice......TO DO.
```
@harisqb @kuriantom369 @tingtingQB Please could you help me complete this here -- it may be simply that they need to come to the Slack organisation to ask a question, but I think if we are saying that something is a little complex, it's necessary to say how to find out more to resolve any issues caused by complexity.

kuriantom369 replied:
Mainly there is a good docs in the pyspark website, the link is mentioned below. It gives a clear idea about various join types, watermarking, and considerations. 

https://spark.apache.org/docs/latest/structured-streaming-programming-guide.html#join-operations

stichbury commented:
Hi @harisqb @kuriantom369 @tingtingQB 

I have now revised the article based on all your feedback. I don't know why GitHub isn't showing the diffs clearly, probably because I changed quite a lot to make things "blog friendly" (that is, i removed bullets in a lot of places because the blog doesn't show them well -- it's not as great at formatting as Brix). 

I took all your comments/suggestions and worked them in. There's one place I could do with some help (marked above) and also would like you to be sure I've got the messaging correct about Kafka. Please could you let me know when you're happy or make any commits to the branch or suggestions here on the PR.

Thank you!

stichbury commented on blog_post_collaboration/spark_streaming_dataset.md:
```diff
@@ -0,0 +1,322 @@
+
+## In summary
+
+In this article, we explained how to use a new Kedro dataset to create streaming pipelines. We created a new Kedro project using the Kedro `pyspark` starter and added two separate Hooks to the Kedro project to enable it to function as a streaming application, then configured the Kedro data catalog to use the new dataset, defining the streaming sources and sinks.
```
Very basic summary as I moved what we had previously up to the top on @harisqb 's suggestion. Is it OK (particularly last section about Kafka)?

kuriantom369 replied:
Kafka part looks fine for me, as we have mentioned why it is not allowed currently :)

tingtingQB commented on blog_post_collaboration/spark_streaming_dataset.md:
```diff
@@ -0,0 +1,323 @@
+
+## Set up a project to integrate Kedro with Spark Structured streaming
+
+The project uses a Kedro dataset to build a structured data pipeline that can read and write data streams with Spark Structured Streaming and process data streams in realtime. You need to add two separate Hooks to the Kedro project to enable it to function as a streaming application. 
```
same here

tingtingQB reviewed (commented):
LGTM! Only 1 minor suggestion!

tingtingQB commented on blog_post_collaboration/spark_streaming_dataset.md:
```diff
@@ -0,0 +1,323 @@
+# A new Kedro dataset for Spark Structured Streaming
+
+This article guides data practitioners on how to set up a Kedro project to integrate Spark Structured Streaming and process data streams in realtime. 
```
closed this as discussed
